Add live data summaries to the homepage

buck54321 opened the pull request:
Add live data summaries to the homepage

This is meant to complement the work being done in PR #921 and be a little experimental, so don't hold back on the criticism. 
    
Adds some additional data summary and display styles to the homepage. Introduces the `threeSigFigs` number formatting functions in both golang template functions and javascript `humanize` helpers, which allows short but precise representations of values from 1e-8 up, using k,M,B prefixes.

In the blocks section, the values of the most recent "proposed" block are displayed prominently, along with an accounting of the current votes for that block. In the mempool section, the count and total DCR for "likely" transactions are displayed, along with a small graphic that illustrates their proportions in the mempool total. "likely" transactions do not include votes for any but the most recently proposed block, or for duplicate votes that can pop up when a stakeholder votes multiple times with a different agenda choice. The "likely" total will count more than 20 new tickets, though.

All new displayed values are shown identically with or without JS enabled in the browser. If JS is enabled, the values will update live. Instead of simply showing the age column when JS is enabled, the age column now replaces the timestamp column, which is redundant. The size column is sacrificially hidden at certain breakpoints so that the tables always display nicely. The vote type column headers are also shortened to a single letter. A new pure-CSS tooltip is used on the headers and elsewhere to add additional information when hovered on desktop or tapped on mobile. Values are shown with the new `threeSigFigs` format, which does mean that users would have to click into the /block or /tx page to get the full 8-point precision.

![image](https://user-images.githubusercontent.com/6109680/51844266-70ab4c00-22da-11e9-8912-53f409c27b0f.png)

Javascript `Mempool` class keeps a very shallow copy of the mempool. No new websocket or API calls are necessary. `Mempool` has methods for retreiving some statistical data about the mempool and tracking votes for recent block(s).

![image](https://user-images.githubusercontent.com/6109680/51844212-4f4a6000-22da-11e9-9ae6-153f2851798f.png)

![image](https://user-images.githubusercontent.com/6109680/51844219-54a7aa80-22da-11e9-9946-25363125e5c1.png)

![image](https://user-images.githubusercontent.com/6109680/51844246-638e5d00-22da-11e9-96fe-cb9583325505.png)

chappjc commented on explorer/explorer_test.go:
```diff
@@ -26,3 +31,18 @@ func TestSimNetName(t *testing.T) {
 	}
 }
+
+// func TestThreeSigFigs(t *testing.T) {
```
Looks like a really nice change in the screenshots.  Is it a WIP still though?  Maybe just this test?

buck54321 commented on views/home.tmpl:
```diff
@@ -186,73 +186,185 @@
+                            <th><div class="position-relative text-center" data-tooltip="tickets">T</div></th>
+                            <th><div class="position-relative text-center" data-tooltip="revocations">X</div></th>
+                            <th><div class="position-relative text-center" data-tooltip="total DCR sent">DCR</div></th>
+                            <th class="d-none d-sm-table-cell d-md-none d-lg-table-cell"><div class="position-relative text-center" data-tooltip="total DCR sent">Size</div></th>
```
oops. wrong tooltip.

chappjc commented on views/home.tmpl:
```diff
@@ -186,73 +186,185 @@
-                    <h4>Latest Blocks</h4> <a href="/blocks" class="pl-2 keyboard-target"><small>see more ...</small></a>
+                <div class="d-flex justify-content-between align-items-center px-2 mx-auto">
+                    <div class="text-left mt-2 pr-3 d-inline-block">
+                      <span class="h6 d-inline-block mb-2 position-relative" data-tooltip="stakeholder's votes for the proposed block.">votes</span><br>
```
"stakeholder's votes for the proposed block"  -> my understanding of what is shown here is that these votes are the ones in mempool that are set to approve/disapprove the current best block (the "proposed" block).  This makes sense since until stakeholders approve the block, the transactions are not considered permanent although they are spendable.

chappjc replied:
NVM, I just saw "proposed" and the block number above it. Very clear now.

chappjc commented on views/home.tmpl:
```diff
@@ -186,73 +186,185 @@
+                      <span class="h6 d-inline-block mb-2 position-relative" data-tooltip="stakeholder's votes for the proposed block.">votes</span><br>
+                      <div class="d-inline-block p1 h4" data-target="homepage.blockVotes" data-hash="{{.Hash}}">
+                        {{range $index, $vote := $.BlockTally}}
+                          <span class="d-inline-block dcricon-{{if eq $vote 0}}reject{{else if eq $vote 1}}affirm{{else}}missing{{end}}"></span>
```
How did you decide to handle votes with the same validator index (same ticket being spent), but with different vote bits set, as is often the case with misconfigured redundant wallets?

chappjc replied:
I think each of `dcricon-{reject,affirm,missing}` deserve tooltips.  Especially in the missing case.  Perhaps for missing it could be something like "this selected ticket has yet to vote" or "waiting for this vote".  The affirm and reject options are much clearer, but I personally think it's worth reiterating whenever we get the chance that the votes is to validate/invalidate the regular transaction tree for this block.

chappjc commented on views/home.tmpl:
```diff
@@ -186,73 +186,185 @@
                         </tr>
                         {{end}}
                     </tbody>
                 </table>
```
I may be in the minority here, but I think that below both the "Latest Blocks" and "Mempool" tables it is worth having a "more" or "see all" kind of link, even though these links exist on the "Latest Blocks" and "Mempool" headers.

chappjc commented on public/scss/home.scss:
```diff
@@ -0,0 +1,73 @@
+}
+
+.block-card {
+  background-color: #efece8;
```
Not sure I love this peach/rust look to the Latest Blocks card.  I don't feel strongly about it though.

chappjc commented on explorer/templates.go:
```diff
@@ -194,6 +194,50 @@ func amountAsDecimalPartsTrimmed(v, numPlaces int64, useCommas bool) []string {
+		return fmt.Sprintf("%.1fk", math.Round(v/100)/10)
+	} else if v >= 1e3 {
+		return fmt.Sprintf("%.2fk", math.Round(v/10)/1e2)
+	} else if v >= 100 {
```
All the cases from here to the end can be covered by the magic of `"g"` where precision applies to the count of all digits, not just the ones to the right of the decimal:

```go
// either of:
strconv.FormatFloat(v, 'g', 3, 64)
fmt.Sprintf("%.3g", v)
```

Note that these functions round rather than truncate so the result should be as intended. https://play.golang.org/p/bgLV1WXNaxZ

buck54321 replied:
There seems to be some important differences. `'%g` will switch to scientific notation if is uses fewer characters. Also, `%g` removes trailing zeros, which can be undesirable.

chappjc replied:
Good point about removing trailing zeros.  However, for the numbers below 100, and sig digits, it won't ever go to scientific notation.

buck54321 replied:
Hmm. Maybe I'm not understanding. https://play.golang.org/p/HUNZjKF-Tui

chappjc replied:
Ah, got it.  Any smaller than 0.0001 and it switches to sci notation.

chappjc commented on explorer/types/explorertypes.go:
```diff
@@ -529,6 +550,80 @@ type VotingInfo struct {
+			c++
+		}
+	}
+	return c
```
Just `return` should be good here.

chappjc commented on explorer/types/explorertypes.go:
```diff
@@ -529,6 +550,80 @@ type VotingInfo struct {
+}
+
+// VoteCount is the number of votes received.
+func (tally *VoteTally) VoteCount() (c int) {
```
Might as well be `VoteCount() int`.

chappjc commented on mempool/collector.go:
```diff
@@ -300,11 +300,15 @@ func ParseTxns(txs []exptypes.MempoolTx, params *chaincfg.Params, lastBlock *Blo
 
-	var totalOut float64
+	var regularTotal, ticketTotal, voteTotal, revTotal float64
+	var likely bool
```
Seems like the meaning is "likely to be mined".  Is that the intended semantics?

buck54321 replied:
More like "likely available to be mined" because it doesn't ever exclude tickets regardless of their count. I'm thinking `likely` -> `mineable` or `likelyMineable`

chappjc replied:
Yeah, either way is good.

chappjc commented on mempool/collector.go:
```diff
@@ -300,11 +300,15 @@ func ParseTxns(txs []exptypes.MempoolTx, params *chaincfg.Params, lastBlock *Blo
+
+	var totalOut, likelyTotal float64
 	var totalSize int32
 	votingInfo := exptypes.VotingInfo{
```
Seems like a `NewVotingInfo` c'tor would be good to have to ensure setting the max votes correctly and making the maps.  BTW, it could set the initial capacity of the maps to params.TicketsPerBlock, but I have a feeling the initial capacity of a default map is above that anyway.

chappjc commented on public/js/helpers/mempool_helper.js:
```diff
@@ -0,0 +1,159 @@
+    }
+  }
+
+  votes (hash) {
```
Would it be fair to describe this function as `tallyVotesOnBlock`?

chappjc commented on public/scss/utils.scss:
```diff
@@ -163,3 +163,32 @@
   white-space: nowrap;
 }
+
+[data-tooltip]:hover::before {
```
I love these `data-tooltip`s!

buck54321 commented:
Works in changes from #921, which included a full rework of the network info column. Switches the columns based on feedback on slack and matrix. Adds a little approved/rejected message when the necessary votes have been received.

Switches the semantics of the vote tallying a little so that a vote result (approved, rejected, not enough votes) is calculated and returned with the slice of votes. Also changed the enum for [no, yes, missing] from 0, 1, 2 to -1, 1, 0, which seemed appropriate for what they represent. Added contructor for `explorertypes.VotingInfo`.

Uses a simple two-tone grey color scheme for now. Changes the `text-success` boostrap class elements to a custom `text-green` because the former had poor contrast on lighter backgrounds and small font size.

![image](https://user-images.githubusercontent.com/6109680/52015121-de0bd800-24a6-11e9-998a-f51df7dd0d06.png)

![image](https://user-images.githubusercontent.com/6109680/52015129-e2d08c00-24a6-11e9-9b69-2737a3094f7e.png)
